dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/filter/GlobalAppShellFilter.java

```diff
@@ -0,0 +1,136 @@
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.webapi.filter;
+
+import static java.util.regex.Pattern.compile;
+
+import java.io.IOException;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import javax.servlet.FilterChain;
+import javax.servlet.RequestDispatcher;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.hisp.dhis.appmanager.AppManager;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+import org.springframework.web.filter.OncePerRequestFilter;
+
+/**
+ * @author Austin McGee <austin@dhis2.org>
+ */
+@Slf4j
+@RequiredArgsConstructor
+@Component
+public class GlobalAppShellFilter extends OncePerRequestFilter {
+  @Autowired AppManager appManager;
+
+  public static final String GLOBAL_APP_SHELL_PATH_PREFIX = "/apps/";
+  public static final String GLOBAL_APP_SHELL_APP_NAME = "global-app-shell";
+
+  private static final Pattern APP_PATH_PATTERN =
+      compile("^/" + "(?:" + AppManager.BUNDLED_APP_PREFIX + "|api/apps/)" + "(\\S+)/(.*)");
+
+  private static final Pattern APP_IN_GLOBAL_SHELL_PATTERN =
+      compile("^" + GLOBAL_APP_SHELL_PATH_PREFIX + "([^/.]+)/?$");
+
+  @Override
+  protected void doFilterInternal(
+      HttpServletRequest request, HttpServletResponse response, FilterChain chain)
+      throws IOException, ServletException {
+    if (!appManager.exists(GLOBAL_APP_SHELL_APP_NAME)) {
+      chain.doFilter(request, response);
+      return;
+    }
+
+    String path = getContextRelativePath(request);
+    if (redirectLegacyAppPaths(request, response, path)) {
+      return;
+    }
+
+    if (path.startsWith(GLOBAL_APP_SHELL_PATH_PREFIX)) {
+      serveGlobalAppShell(request, response, path);
+      return;
+    }
+
+    chain.doFilter(request, response);
+  }
+
+  private boolean redirectLegacyAppPaths(
+      HttpServletRequest request, HttpServletResponse response, String path) throws IOException {
+    String queryString = request.getQueryString();
+    Matcher m = APP_PATH_PATTERN.matcher(path);
+
+    boolean matchesPattern = m.find();
+    boolean isIndexPath = path.endsWith("/") || path.endsWith("/index.html");
+    boolean hasRedirectFalse = queryString == null || !queryString.contains("redirect=false");
+    if (matchesPattern && isIndexPath && hasRedirectFalse) {
+      String appName = m.group(1);
+      response.sendRedirect(request.getContextPath() + GLOBAL_APP_SHELL_PATH_PREFIX + appName);
```

Comment:
Also, if redirect=false, should we redirect?

Reply:
Maybe its just the name of the variable that confuses me

Reply:
Variable name was incorrect, fixed now